[GeoMechanicsApplication] Add Coulomb model with tension cutoff for interface elements

mnabideltares opened the pull request:
**📝 Description**
The aim of this issue is to have a constitutive law that is applicable to (line) interface elements, which adopts the Coulomb model.

The idea here is to split the functionality such that parts of it can be reused by the Mohr-Coulomb model for continuum elements.

We can for instance look at some available (Fortran) implementation.

The common part includes the return mapping algorithm (see example implementation for the details):
- Calculate elastic trial stress: $`\sigma_{1} = \sigma_{0} + D_{\mathrm{el}} \Delta \epsilon`$
- Evaluation of both yield functions:
  1. shear yield surface
  2. tension cut-off
- Calculate the return direction.
- Calculate the plastic increment.
- Calculate the constitutive matrix.

The element-specific (i.e. interface or continuum element) part include:
- Prediction of the elastic trial stress based on the strain (or relative displacement) increment.
- Calculate the principal stresses and principal directions.

**🆕 Changelog**
- Refactored the continum Mohr-Coulomb to be generic
- Added Mohr-Coulomb with tension cutoff for the interface element
- Registered the law in Kratos GeoMechanicsApplication
- Added unit tests
- Added integration test

mnabideltares committed:
Shifted Mohr_Coulomb to unitlities

mnabideltares committed:
gradual modification to sigma-tau replacing of principal stresses

mnabideltares committed:
intermediate development

mnabideltares committed:
Refactored Mohr-Coulomb

mnabideltares committed:
Added Mohr-Coulomb for interface, registered the law

mnabideltares committed:
Added unit tests

mnabideltares committed:
Added unit test + fix a bug

mnabideltares committed:
Modified the doc about the constitutive law of Mohr-Coulomb

mnabideltares committed:
Added documentation for the integration test

mnabideltares committed:
removed files to rename , will be followed by next commit

mnabideltares committed:
renamed the files and added again

mnabideltares committed:
corrected figure

mnabideltares committed:
renaming files to lowercase

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.cpp:
```diff
@@ -25,17 +25,15 @@ CoulombYieldSurface::CoulombYieldSurface(double FrictionAngleInRad, double Cohes
 }
 
-Vector CoulombYieldSurface::DerivativeOfFlowFunction(const Vector& rPrincipalStress) const
+Vector CoulombYieldSurface::DerivativeOfFlowFunction(const Vector& rSigmaTau) const
```
As far as I understand this is only for geo_mechanics, if yes, can we change it signature to ()? rSigmaTau is not used in this function.

avdg81 replied:
This is an implementation of the yield surface interface, where in general you'd like to receive the current stress state. Since that stress state is irrelevant for this particular implementation, I'd suggest to "unname" the function parameter:
```suggestion
Vector CoulombYieldSurface::DerivativeOfFlowFunction(const Vector&) const
```
Naturally, this should also be done in the corresponding header file.

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,222 @@
+    return p_result;
+}
+
+Vector& InterfaceMohrCoulombWithTensionCutOff::GetValue(const Variable<Vector>& rThisVariable, Vector& rValue)
```
Please rename `rThisVariable `to `rVariable`

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,222 @@
+        << "value of " << rVariable.Name() << " for property " << rMaterialProperties.Id()
+        << " is out of range: " << rMaterialProperties[rVariable] << " is not in [0.0, "
+        << (MaxValue ? std::to_string(*MaxValue) + "]" : "->") << std::endl;
+}
```
This is an exact copy of the function from mohr_coulomb_with_ternsion-cutoff.cpp. Is it possible to re-used it?

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,222 @@
+}
+
+Matrix& InterfaceMohrCoulombWithTensionCutOff::CalculateValue(ConstitutiveLaw::Parameters& rParameterValues,
+                                                              const Variable<Matrix>& rThisVariable,
```
Please rename `rThisVariable `to `rVariable`

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_mohr_coulomb_with_tension_cutoff.h:
```diff
@@ -0,0 +1,73 @@
+    void                                   InitializeMaterial(const Properties&     rMaterialProperties,
+                                                              const Geometry<Node>& rElementGeometry,
+                                                              const Vector&         rShapeFunctionsValues) override;
+    void    InitializeMaterialResponseCauchy(Parameters& rValues) override;
```
Is it better to use `Parameters& rParameters` as in other places?

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_mohr_coulomb_with_tension_cutoff.h:
```diff
@@ -0,0 +1,73 @@
+                                                              const Geometry<Node>& rElementGeometry,
+                                                              const Vector&         rShapeFunctionsValues) override;
+    void    InitializeMaterialResponseCauchy(Parameters& rValues) override;
+    Vector& GetValue(const Variable<Vector>& rThisVariable, Vector& rValue) override;
```
Please rename `rThisVariable `as `rVariable`

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -0,0 +1,222 @@
+        HasInitialState() ? GetInitialState().GetInitialStressVector() : ZeroVector{GetStrainSize()};
+}
+
+void InterfaceMohrCoulombWithTensionCutOff::InitializeMaterialResponseCauchy(Parameters& rValues)
```
Is it better to use `Parameters& rParameters` as in other places?

avdg81 replied:
As discussed, it has been renamed to `rConstitutiveLawParameters`.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_mohr_coulomb_with_tension_cutoff.h:
```diff
@@ -0,0 +1,73 @@
+                            const ProcessInfo&  rCurrentProcessInfo) const override;
+    void CalculateMaterialResponseCauchy(ConstitutiveLaw::Parameters& rParameters) override;
+    void FinalizeMaterialResponseCauchy(ConstitutiveLaw::Parameters& rValues) override;
+    Matrix& CalculateValue(Parameters& rParameterValues, const Variable<Matrix>& rThisVariable, Matrix& rValue) override;
```
Please rename `rThisVariable `as `rVariable`

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -36,6 +36,85 @@ double GetValueOfUMatParameter(const Properties& rProperties, const Variable<int
+    return result;
+}
+
+Vector ReturnStressAtTensionApexReturnZone(const Vector& rSigmaTau, double TensileStrength)
```
This function is called only in one place. Is it better to remove argument `const Vector& rSigmaTau`?

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -36,6 +36,85 @@ double GetValueOfUMatParameter(const Properties& rProperties, const Variable<int
+    return rSigmaTau + lambda * rDerivativeOfFlowFunction;
+}
+
+Vector ReturnStressAtCornerReturnZone(const Vector& rSigmaTau, const Vector& rCornerPoint)
```
This function is called only in one place. Is it better to remove argument const Vector& rSigmaTau?
Secondly, the name `rCornerPoint `is confusing. It is calculated with `CalculateCornerPoint `function and here it is returned as a stress vector. Is it the location vector or the stress vector?

avdg81 replied:
You are correct that it was simply returning a copy of the corner point (which represents a particular stress state). As such, we have eliminated this function, and use the corner point directly.

markelov208 commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -36,6 +36,85 @@ double GetValueOfUMatParameter(const Properties& rProperties, const Variable<int
+    const auto cof1      = std::sin(FrictionAngle);
+    const auto cof2      = Cohesion * std::cos(FrictionAngle);
+    const auto numerator = cof1 * rDerivativeOfFlowFunction[0] + rDerivativeOfFlowFunction[1];
+    const auto lambda    = (cof2 - rSigmaTau[0] * cof1 - rSigmaTau[1]) / numerator;
```
`cof2 `is used only once, it is possible to merge these two lines. Then `coef1 `can be just `coefficient`.

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.cpp:
```diff
@@ -80,4 +159,44 @@ double ConstitutiveLawUtilities::GetFrictionAngleInDegrees(const Properties& rPr
+    const auto     tension_yield_function = rTensionCutOff.YieldFunctionValue(rSigmaTau);
+    const auto     coulomb_tolerance      = tolerance * (1.0 + std::abs(coulomb_yield_function));
+    const auto     tension_tolerance      = tolerance * (1.0 + std::abs(tension_yield_function));
+    return coulomb_yield_function < coulomb_tolerance && tension_yield_function < tension_tolerance;
```
If `coulomb_yield_function `and `tension_yield_function `are always positive and there is no need to take `abs `for tolerance or in line 199 `abs `shall be used for these variables as well.

avdg81 replied:
Yield function values can also be negative (i.e. when in the elastic region).

markelov208 commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.h:
```diff
@@ -35,6 +37,13 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) ConstitutiveLawUtilities
     static double GetCohesion(const Properties& rProperties);
     static double GetFrictionAngleInDegrees(const Properties& rProperties);
+    static Vector MapStressesInMorhCoulomb(const Properties&          rProperties,
+                                           Vector&                    rSigmaTau,
```
Please add const

avdg81 replied:
This function has been removed.

markelov208 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -36,13 +44,16 @@ KRATOS_TEST_CASE_IN_SUITE(TestCoulombYieldSurface, KratosGeoMechanicsFastSuiteWi
     Vector principal_stress(3);
     principal_stress <<= 3.0, 2.0, 1.0;
-    KRATOS_EXPECT_NEAR(coulomb_yield_surface.YieldFunctionValue(principal_stress), 1.0, Defaults::absolute_tolerance);
+    Vector sigma_tau = ConvertPrincipalStressesToSigmaTau(principal_stress);
```
Please change `Vector `to `auto`

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_yield_surface.cpp:
```diff
@@ -26,6 +26,14 @@ using namespace std::string_literals;
+    result(1) = 0.5 * (rPrincipalStresses(0) - rPrincipalStresses(2));
+    return result;
+}
+
```
For me it is not clear why is a principal stress in this file? I would name this function as CreateSigmaTauVector with two doubles as the input.

WPK4FEM replied:
Principal stresses are the eigenvalues of a stress tensor. The definition of the sigma - tau space is in terms of principal stresses. Please keep it like this.

avdg81 replied:
Indeed, this is a copy of `StressStrainUtilities::TransformPrincipalStressesToSigmaTau`. Better use that here, and remove the local copy.

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/tests/test_mohr_coulomb_with_tension_cutoff/README.md:
```diff
@@ -23,4 +24,13 @@ The boundary conditions are shown below:
+
+<img src="interface_element.svg" width="600">
+
+An increasing normal load of 600 Pa is applied over 2 seconds. This is followed by an increasing compressive deformation of 0.1 m in the $\tau$ direction, applied over the next 2 seconds.
```
Does it mean that the load is increased from 0 up to 600 Pa in two seconds?

avdg81 replied:
Yes, that is what I understand when I read the corresponding table:
```text
Begin Table 2 TIME NORMAL_LOAD
  0.0  0.0
  2.0  600.0
  4.0  600.0
End Table
```

So, the normal load increases linearly from 0 at $`t = 0`$ to 600 at $`t = 2`$. For $`t > 2`$, the normal load remains at 600.

markelov208 replied:
Thank you. I would suggest to use this sentence in the description.

avdg81 replied:
Done.

markelov208 reviewed (changes requested):
Hi Mohamed, thank you very much for this new functionality. I have a number of small comments. As well two more unit tests can be added for Check and Clone functions as it is done for the Mohr-Coulomb.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.h:
```diff
@@ -28,7 +28,7 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) CoulombYieldSurface : public YieldSu
 
-    [[nodiscard]] double YieldFunctionValue(const Vector& rPrincipalStress) const override;
+    [[nodiscard]] double YieldFunctionValue(const Vector& rSigmaTau) const override;
     [[nodiscard]] Vector DerivativeOfFlowFunction(const Vector& rPrincipalStress) const override;
```
As per the previous comment, let's "unname" this function parameter:
```suggestion
    [[nodiscard]] Vector DerivativeOfFlowFunction(const Vector&) const override;
```

avdg81 replied:
Done.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/146-mohr-coulomb-for-interface

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -47,59 +48,10 @@ Vector TransformPrincipalStressesToSigmaAndTau(const Vector& rPrincipalStresses)
 }
 
-double CalculateApex(double FrictionAngle, double Cohesion)
+void TransformSigmaAndTauToPrincipalStresses(const Vector& rSigmaTau, Vector& rPrincipalStresses)
```
Shouldn't this function be part of the stress utilities? We should at least keep it close to the inverse operation.

avdg81 committed:
Moved stress transformations to utility functions

avdg81 committed:
Forgotten to reformat the code

avdg81 committed:
Removed several unused functions

avdg81 committed:
Renamed two member functions

avdg81 committed:
Removed erroneous "Morhr"

avdg81 committed:
Use "traction" and "relative displacement" for interface law

avdg81 committed:
Don't reimplement the calculation of the constitutive matrix

Also renamed a few function parameters.

avdg81 committed:
Introduced a dummy class for Coulomb with a tension cut-off

It will be populated in upcoming commits. The intention is to have a single logical place for Coulomb-related functionality.

avdg81 committed:
Added constructors to the new implementation class

avdg81 committed:
Use data member to check for admissible stress state

Also added (de)serialization for the new implementation class.

avdg81 committed:
Use data member to do the return mapping

avdg81 committed:
Fixed the sign of the shear traction after return mapping

avdg81 committed:
Renamed a test directory

Also removed an empty input data structure.

avdg81 committed:
Renamed a few files

Also renamed a few functions.

avdg81 committed:
Removed data members that were no longer used

Also moved two `#include`s from header files to implementation files.

Fixed the `Clone` member functions, which did not copy the new implementation class objects.

avdg81 committed:
Use constitutive law utility functions

avdg81 committed:
Removed an unused function parameter

avdg81 committed:
Avoid hiding a member function of a base class

avdg81 committed:
Renamed three function parameters

avdg81 committed:
Use initial state rather than finalizing before doing the calculation

avdg81 committed:
Improved readability of the test code

avdg81 committed:
Improved readability of the second unit test

avdg81 committed:
Improved readability of the third unit test

avdg81 committed:
Improved readability of the fourth unit test

avdg81 committed:
Improved readability of the fifth unit test

avdg81 committed:
Removed an unused helper function

avdg81 committed:
Added a unit test for (de)serialization of the Coulomb interface law

avdg81 committed:
Improved the return value of `WorkingSpaceDimension`

avdg81 committed:
Removed "Mohr" from interface-related names

Also renamed a unit test file.

avdg81 committed:
Added a unit test for member `Check`

avdg81 committed:
Added a unit test for member `GetStressMeasure`

avdg81 committed:
Added four small unit tests for class `InterfaceCoulombWithTensionCutOff`

avdg81 committed:
Added a unit test for member `Clone`

avdg81 committed:
Extended the unit tests

- Added a new unit test for the calculation of the constitutive matrix.
- Extended the unit test for the elastic state using reversed shear.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/146-mohr-coulomb-for-interface

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -0,0 +1,163 @@
+
+using namespace Kratos;
+
+double CalculateApex(double FrictionAngle, double Cohesion)
```
Just for for clarity:
```suggestion
double CalculateApex(double FrictionAngleInRadians, double Cohesion)
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -0,0 +1,163 @@
+    return Cohesion / std::tan(FrictionAngle);
+}
+
+Vector CalculateCornerPoint(double FrictionAngle, double Cohesion, double TensileStrength)
```
```suggestion
Vector CalculateCornerPoint(double FrictionAngleInRadians, double Cohesion, double TensileStrength)
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -0,0 +1,163 @@
+           rCornerPoint[1] - rTrialSigmaTau[1] - rCornerPoint[0] + rTrialSigmaTau[0] > 0.0;
+}
+
+bool IsStressAtCornerReturnZone(const Vector& rTrialSigmaTau, double DilatancyAngle, const Vector& rCornerPoint)
```
```suggestion
bool IsStressAtCornerReturnZone(const Vector& rTrialSigmaTau, double DilatancyAngleInRadians, const Vector& rCornerPoint)
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -0,0 +1,163 @@
+CoulombWithTensionCutOffImpl::CoulombWithTensionCutOffImpl(double FrictionAngleInRad,
+                                                           double Cohesion,
+                                                           double DilatationAngleInRad,
+                                                           double TensileStrength)
```
Some renaming suggestions:
```suggestion
CoulombWithTensionCutOffImpl::CoulombWithTensionCutOffImpl(double FrictionAngleInRadians,
                                                           double Cohesion,
                                                           double DilatancyAngleInRadians,
                                                           double TensileStrength)
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -0,0 +1,163 @@
+
+Vector CoulombWithTensionCutOffImpl::DoReturnMapping(const Properties& rProperties, const Vector& rTrialSigmaTau) const
+{
+    Vector result;
```
Remove this unused local variable.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -0,0 +1,163 @@
+                                    ConstitutiveLawUtilities::GetCohesion(rProperties));
+    const auto corner_point = CalculateCornerPoint(
+        ConstitutiveLawUtilities::GetFrictionAngleInRadians(rProperties),
+        ConstitutiveLawUtilities::GetCohesion(rProperties), rProperties[GEO_TENSILE_STRENGTH]);
```
Move this variable declaration and initialization closer to its first usage.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.h:
```diff
@@ -0,0 +1,44 @@
+namespace Kratos
+{
+
+class Properties;
```
Let's also explicitly forward declare `class Serializer;` here. We now get it indirectly through one of the `#include`s, but I feel it's better to be explicit about it.
```suggestion
class Properties;
class Serializer;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.h:
```diff
@@ -0,0 +1,44 @@
+{
+public:
+    CoulombWithTensionCutOffImpl() = default;
+    CoulombWithTensionCutOffImpl(double FrictionAngleInRad, double Cohesion, double DilatationAngleInRad, double TensileStrength);
```
```suggestion
    CoulombWithTensionCutOffImpl(double FrictionAngleInRadians, double Cohesion, double DilatancyAngleInRadians, double TensileStrength);
```

avdg81 committed:
"Unnamed" an unused function parameter

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.h:
```diff
@@ -0,0 +1,68 @@
+namespace Kratos
+{
+
+class ConstitutiveLawDimension;
```
This forward declaration is not being used, so let's remove it.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+        mRelativeDisplacementVectorFinalized = rValues.GetStrainVector();
+        mIsModelInitialized                  = true;
+    }
+}
```
Not sure whether we actually need this member function. It is not covered by any the unit tests. Probably using the initial state is the way to go (see member function `InitializeMaterial`).

When we decide to remove this member function, make sure to let member `RequiresInitializeMaterialResponse` return `false`. We can then also remove data member `mIsModelInitialized`.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+
+// Application includes
+#include "custom_constitutive/interface_coulomb_with_tension_cut_off.h"
+#include "custom_constitutive/constitutive_law_dimension.h"
```
This `#include` is not being used here.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+        << "value of " << rVariable.Name() << " for property " << rMaterialProperties.Id()
+        << " is out of range: " << rMaterialProperties[rVariable] << " is not in [0.0, "
+        << (MaxValue ? std::to_string(*MaxValue) + "]" : "->") << std::endl;
+}
```
This local function is an exact copy of the one found in `mohr_coulomb_with_tension_cutoff.cpp`. I'd suggest to make it a `static` member of class `ConstitutiveLawUtilities`.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+
+namespace Kratos
+{
+ConstitutiveLaw::Pointer InterfaceCoulombWithTensionCutOff::Clone() const
```
As it stands now, this member lacks copying data member `mIsModelInitialized`. However, if we decide to remove member function `InitializeMaterialResponseCauchy` (and thus remove data member `mIsModelInitialized`), this correction no longer applies.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+    return p_result;
+}
+
+Vector& InterfaceCoulombWithTensionCutOff::GetValue(const Variable<Vector>& rVariable, Vector& rValue)
+{
+    if (rVariable == CAUCHY_STRESS_VECTOR) {
+        rValue = mTractionVector;
+    } else {
+        rValue = ConstitutiveLaw::GetValue(rVariable, rValue);
+    }
+    return rValue;
+}
```
I doubt whether we really need this member function. So far, it seems to be mostly (if not exclusively) used by unit tests.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+    return rValue;
+}
+
+void InterfaceCoulombWithTensionCutOff::SetValue(const Variable<Vector>& rVariable,
+                                                 const Vector&           rValue,
+                                                 const ProcessInfo&      rCurrentProcessInfo)
+{
+    if (rVariable == CAUCHY_STRESS_VECTOR) {
+        mTractionVector = rValue;
+    } else {
+        KRATOS_ERROR << "Can't set value of " << rVariable.Name() << ": unsupported variable\n";
+    }
+}
```
I'm also hesitant about this member function. Do we really need it? When there is an initial state to take into account, we should use `SetInitialState`, `GetInitialState`, and `HasInitialState` (it's part of any constitutive law).

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+    }
+}
+
+SizeType InterfaceCoulombWithTensionCutOff::WorkingSpaceDimension() { return N_DIM_2D; }
```
Let's add a comment here:
```suggestion
SizeType InterfaceCoulombWithTensionCutOff::WorkingSpaceDimension()
{
    // Note that this implementation assumes line interface elements. It needs to be modified when planar interface elements become available.
    return N_DIM_2D;
}
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+
+SizeType InterfaceCoulombWithTensionCutOff::GetStrainSize() const
+{
+    return VOIGT_SIZE_2D_INTERFACE;
```
Let's add a comment here, too:
```suggestion
    // Note that this implementation assumes line interface elements. It needs to be modified when planar interface elements become available.
    return VOIGT_SIZE_2D_INTERFACE;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+
+void InterfaceCoulombWithTensionCutOff::CalculateMaterialResponseCauchy(ConstitutiveLaw::Parameters& rParameters)
+{
+    const auto& r_prop = rParameters.GetMaterialProperties();
```
```suggestion
    const auto& r_properties = rParameters.GetMaterialProperties();
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+
+    if (negative) trial_sigma_tau[1] *= -1.0;
+
+    mTractionVector = trial_sigma_tau;
```
I'm not sure whether we actually need `mTractionVector`. Note that we **don't** have a similar data member for the current relative displacement vector. For that, we rely on the value stored in `ConstitutiveLaw::Parameters`. For instance, see member function `FinalizeMaterialResponseCauchy`. Perhaps we can apply the same philosophy to the current traction vector?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+void InterfaceCoulombWithTensionCutOff::FinalizeMaterialResponseCauchy(ConstitutiveLaw::Parameters& rValues)
+{
+    mRelativeDisplacementVectorFinalized = rValues.GetStrainVector();
+    mTractionVectorFinalized             = mTractionVector;
```
Is there any reason why we don't use `rValues.GetStressVector()`? That would be more consistent with how we handle relative displacements...

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+        rValue                   = MakeConstitutiveMatrix(r_properties[INTERFACE_NORMAL_STIFFNESS],
+                                                          r_properties[INTERFACE_SHEAR_STIFFNESS]);
+    } else {
+        KRATOS_ERROR << "Can't calculate value of " << rVariable.Name() << ": unsupported variable\n";
```
Shouldn't we call the base class version of this member after we have handled the specific cases?

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,219 @@
+    result(0, 0) = NormalStiffness;
+    result(1, 1) = ShearStiffness;
+    return result;
+}
```
This is an exact copy of `GeoIncrementalLinearElasticInterfaceLaw::MakeConstitutiveMatrix`. I'd suggest to move this to a common place to avoid duplication.

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/tension_cutoff.cpp:
```diff
@@ -20,15 +20,15 @@ namespace Kratos
 }
 
-Vector TensionCutoff::DerivativeOfFlowFunction(const Vector& rPrincipalStress) const
+Vector TensionCutoff::DerivativeOfFlowFunction(const Vector& rSigmaTau) const
```
Since we don't use this function parameter, let's "unname" it:
```suggestion
Vector TensionCutoff::DerivativeOfFlowFunction(const Vector&) const
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/tension_cutoff.h:
```diff
@@ -28,7 +28,7 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) TensionCutoff : public YieldSurface
 
-    [[nodiscard]] double YieldFunctionValue(const Vector& rPrincipalStress) const override;
+    [[nodiscard]] double YieldFunctionValue(const Vector& rSigmaTau) const override;
     [[nodiscard]] Vector DerivativeOfFlowFunction(const Vector& rPrincipalStress) const override;
```
And "unname" it here too:
```suggestion
    [[nodiscard]] Vector DerivativeOfFlowFunction(const Vector&) const override;
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/yield_surface.h:
```diff
@@ -25,7 +25,7 @@ class YieldSurface
 
-    [[nodiscard]] virtual double YieldFunctionValue(const Vector& rPrincipalStress) const       = 0;
+    [[nodiscard]] virtual double YieldFunctionValue(const Vector& rSigmaTau) const              = 0;
     [[nodiscard]] virtual Vector DerivativeOfFlowFunction(const Vector& rPrincipalStress) const = 0;
```
Shouldn't this be `rSigmaTau` too?

avdg81 commented on applications/GeoMechanicsApplication/custom_utilities/constitutive_law_utilities.h:
```diff
@@ -15,6 +15,8 @@
 // Project includes
 #include "containers/variable.h"
+#include "custom_constitutive/coulomb_yield_surface.h"
+#include "custom_constitutive/tension_cutoff.h"
```
These `#include`s are probably no longer needed here...

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,456 @@
+//
+
+#include "custom_constitutive/interface_coulomb_with_tension_cut_off.h"
+#include "custom_constitutive/plane_strain.h"
```
This `#include` is not being used here.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,456 @@
+    // Arrange: set elastic tensile state
+    traction_vector <<= 5.0, 4.0;
+    const auto dummy_process_info = ProcessInfo{};
+    law.SetValue(CAUCHY_STRESS_VECTOR, traction_vector, dummy_process_info);
```
We can get rid of a local variable by using a temporary object:
```suggestion
    law.SetValue(CAUCHY_STRESS_VECTOR, traction_vector, ProcessInfo{});
```
In terms of readability, in my opinion, both ways are similar. The suggestion is, however, simpler.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,456 @@
+    KRATOS_EXPECT_VECTOR_EQ(parameters.GetStressVector(), calculated_traction_vector);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(InterfaceCoulombWithTensionCutOff_WorkingSpaceDimensionIsAlways2D,
```
```suggestion
KRATOS_TEST_CASE_IN_SUITE(InterfaceCoulombWithTensionCutOff_Has2DWorkingSpaceDimension,
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,456 @@
+    KRATOS_EXPECT_EQ(InterfaceCoulombWithTensionCutOff{}.WorkingSpaceDimension(), N_DIM_2D);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(InterfaceCoulombWithTensionCutOff_StressMeasureIsAlwaysCauchy,
```
```suggestion
KRATOS_TEST_CASE_IN_SUITE(InterfaceCoulombWithTensionCutOff_HasCauchyStressMeasure,
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,456 @@
+    KRATOS_EXPECT_EQ(InterfaceCoulombWithTensionCutOff{}.GetStressMeasure(), ConstitutiveLaw::StressMeasure_Cauchy);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(InterfaceCoulombWithTensionCutOff_StrainSizeIsAlways2, KratosGeoMechanicsFastSuiteWithoutKernel)
```
```suggestion
KRATOS_TEST_CASE_IN_SUITE(InterfaceCoulombWithTensionCutOff_StrainSizeEqualsTwo, KratosGeoMechanicsFastSuiteWithoutKernel)
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,456 @@
+    KRATOS_EXPECT_EQ(InterfaceCoulombWithTensionCutOff{}.GetStrainSize(), 2);
+}
+
+KRATOS_TEST_CASE_IN_SUITE(InterfaceCoulombWithTensionCutOff_StrainMeasureIsAlwaysInfinitesimal,
```
```suggestion
KRATOS_TEST_CASE_IN_SUITE(InterfaceCoulombWithTensionCutOff_HasInfinitesimalStrainMeasure,
```

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -504,4 +504,134 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_Get3DLawFeatures, KratosG
 
-} // namespace Kratos::Testing
\ No newline at end of file
+KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_CalculateMaterialResponseCauchyAtElasticZoneInterface,
```
This is not about the interface formulation. This comment also holds for the next four unit tests.

avdg81 committed:
Don't duplicate a utility function

avdg81 committed:
Prefer `auto` (or `const auto`) if possible

avdg81 committed:
Removed a redundant function

avdg81 committed:
Renamed a few function parameters

Also removed an unused local variable and simplified a vector construction.

avdg81 committed:
Added a missing forward declaration

avdg81 committed:
Moved a duplicated utility function

Also removed a few redundant `#include`s.

avdg81 committed:
Moved a duplicated utility function

Other changes include:
- Renamed a local variable.
- Added two comments.

avdg81 committed:
Don't repeat the calculation of the first principal stress component

Other changes include:
- Removed an unused forward declaration.
- "Unnamed" an unused function parameter.
- Simplified the construction of a vector with identical values.

avdg81 committed:
Renamed a function parameter

avdg81 commented on applications/GeoMechanicsApplication/tests/test_mohr_coulomb_with_tension_cutoff/interface_element.svg:
I believe $`\tau`$ should be $`t`$.

avdg81 committed:
Several minor changes

- Removed an unnecessary `#include`.
- Eliminated a redundant local variable.
- Renamed four unit test cases.

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/146-mohr-coulomb-for-interface

avdg81 reviewed (commented):
Several suggestions, must of them being (very) minor in nature. We're almost there...

avdg81 committed:
Fixed an extra qualification on member error

Also reformatted the code using Clang-format.

avdg81 committed:
Processed more review suggestions

- Renamed a few function parameters.
- Removed several redundant name qualifiers.
- Eliminated a local variable and renamed another one.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -0,0 +1,157 @@
+
+double CalculateApex(double FrictionAngleInRadians, double Cohesion)
+{
+    return Cohesion / std::tan(FrictionAngleInRadians);
```
As far as I understand the friction angle has to be larger than 0. However, it is checked only for non-negative values. Should the check to be extanded ]0;90[?

avdg81 replied:
No, the friction angle can be 0.0. In fact, the Tresca criterion uses a non-zero cohesion in combination with a friction angle of 0.0.

markelov208 replied:
Then here is a division by zero in case of zero friction angle. Is it OK that this function will return NaN?

mnabideltares replied:
I think the friction angle of zero has no physical meaning and the user must not be allowed to set it as zero. Maybe can avoid the zero value in the check?

avdg81 replied:
As discussed, we have set the bounds of $`\phi`$ to $`(0, 90)`$.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -0,0 +1,157 @@
+
+namespace Kratos
+{
+CoulombWithTensionCutOffImpl::CoulombWithTensionCutOffImpl(double FrictionAngleInRadians,
```
I think `Impl `means `implicit `because WPK once told it. Would it be useful to use `Implicit`?

avdg81 replied:
Here, `Impl` is short for `Implementation`. It is a quite common abbreviation (think for instance of the `pImpl` idiom), so I would suggest to keep it the way it is now.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/incremental_linear_elastic_interface_law.cpp:
```diff
@@ -53,8 +54,8 @@ Matrix& GeoIncrementalLinearElasticInterfaceLaw::CalculateValue(ConstitutiveLaw:
 {
     if (rThisVariable == CONSTITUTIVE_MATRIX) {
```
There are still a number of `rThisVariable`

avdg81 replied:
Nice suggestion. However, we feel that this is better done in a separate PR, to avoid "scope creep". This file (that is about the linear elastic law for interface elements) is not the primary focus of the current PR.

markelov208 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_interface_coulomb_with_tension_cut_off.cpp:
```diff
@@ -0,0 +1,453 @@
+
+    // Assert
+    auto expected_traction_vector = Vector{2};
+    expected_traction_vector <<= -6.0, 8.0;
```
Would it be better `expected_traction_vector` = `initial_traction_vector`?

avdg81 replied:
Done.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -0,0 +1,157 @@
+{
+}
+
+bool CoulombWithTensionCutOffImpl::IsAdmissibleSigmaTau(const Vector& rTrialSigmaTau) const
```
As far as I understand `amissible `is a common keyword for Mohr-Coulomb theory. However, would it be clearer to use `IsElasticRegion`?

avdg81 replied:
The "admissible" stress states also include ones where $`F = 0`$ (i.e. when plasticity occurs). So we'd rather keep the name as-is.

markelov208 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -504,4 +504,134 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_Get3DLawFeatures, KratosG
+    Vector cauchy_stress_vector(4);
+    cauchy_stress_vector <<= -6.0, 0.0, 0.0, 8.0;
+    Vector expected_cauchy_stress_vector(4);
+    expected_cauchy_stress_vector <<= -6.0, 0.0, 0.0, 8.0;
```
as above, should  `expected_cauchy_stress_vector `= `cauchy_stress_vector`?

avdg81 replied:
Done.

markelov208 reviewed (commented):
Hi Anne, thank you very much for the massive work. I have a few non-blocking comments.

avdg81 committed:
The expected elastic traction vectors do not need to be repeated

avdg81 committed:
Updated the test documentation

mnabideltares committed:
figure is modified

markelov208 reviewed (dismissed):
Hi Anne, thank you very much for the clarifications. I still have the non-blocking comment on the zeroth friction angle which leads to a division by zero. Probably, it is handled later in the code properly, then should a unit test be added for this value?

avdg81 committed:
Avoid division by zero

The friction angle must be in the range (0.0, 90.0).

markelov208 reviewed (approved):
I think the PR is ready to be merged.
